### ISSUE TITLE
Feature: Implement Neural Network Genome ID Embeddings

### ISSUE THREAD

**Issue by csmangum** [Dooders/AgentFarm#90]
Create a neural network-based embedding system for genome IDs that can learn meaningful representations of our genetic space. This will allow us to efficiently compare genomes, find similarities, and potentially predict fitness or behavior.

## Objectives
- Create a neural encoder that converts genome IDs into fixed-dimension embeddings
- Implement training infrastructure for learning meaningful embeddings
- Provide utilities for genome comparison and similarity search
- Integrate with existing genome ID system

#### GenomeEncoder Class
```python
class GenomeEncoder(nn.Module):
    def __init__(self, embedding_dim: int = 32):
        # Embedding layers for:
        # - generation number (discrete embedding)
        # - parent hash (hex to binary to embedding)
        # - trait hash (hex to binary to embedding)
        # Final FC layers to embedding space
```

#### Enhanced GenomeId Class
```python
class GenomeId(BaseModel):
    # Existing fields plus embedding support
    # Methods for converting to/from embeddings
```

#### Phase 1: Basic Structure
- [ ] Create new module `genome_embeddings/`
- [ ] Implement basic `GenomeEncoder` architecture
- [ ] Add embedding support to `GenomeId` class
- [ ] Write tests for basic functionality

#### Phase 2: Training Infrastructure
- [ ] Create `GenomeDataset` class
- [ ] Implement contrastive loss function
- [ ] Create training loop with:
  - Positive pair sampling (same family/generation)
  - Negative pair sampling (different families)
  - Batch processing
  - Validation metrics

#### Phase 3: Utility Functions
- [ ] Implement similarity search
- [ ] Add visualization tools (t-SNE/UMAP)
- [ ] Create genome clustering utilities
- [ ] Add serialization/deserialization support

#### Phase 4: Integration
- [ ] Integrate with existing agent repository
- [ ] Add database storage for embeddings
- [ ] Create API endpoints for embedding operations
- [ ] Update documentation

### Dependencies to Add
```toml
torch>=2.0.0
numpy>=1.20.0
scikit-learn>=1.0.0  # for t-SNE
umap-learn>=0.5.0    # for UMAP visualization
```

### New Files Needed
```
genome_embeddings/
├── __init__.py
├── encoder.py          # GenomeEncoder implementation
├── dataset.py          # Training dataset
├── training.py         # Training loop and loss functions
├── visualization.py    # Plotting and visualization
└── utils.py           # Utility functions
```

### Database Changes
- Add embedding vector storage
- Add index for similarity search

## Testing Requirements
- [ ] Unit tests for encoder architecture
- [ ] Integration tests for training pipeline
- [ ] Performance tests for similarity search
- [ ] Validation tests for embedding quality

## Acceptance Criteria
1. Encoder can convert any valid genome ID to fixed-dimension embedding
2. Similar genomes (same family/generation) have closer embeddings than dissimilar ones
3. Can find similar genomes in < 100ms for database of 10k genomes
4. Visualization tools show clear clustering of related genomes
5. All tests pass with >90% coverage
6. Documentation includes:
   - Architecture overview
   - Training instructions
   - Usage examples
   - Performance benchmarks

## Future Enhancements
- Fine-tuning capabilities
- Support for different embedding dimensions
- Online learning for continuous adaptation
- Multi-GPU training support
- Export to ONNX format

## Resources
- [Contrastive Learning Paper](https://arxiv.org/abs/2002.05709)
- [SimCLR Framework](https://arxiv.org/abs/2002.05709)
- [Efficient Similarity Search](https://arxiv.org/abs/1603.09320)

## Notes
- Consider using FAISS for efficient similarity search in production
- May want to experiment with different architectures for the encoder
- Should benchmark memory usage for large genome databases
- Consider adding periodic retraining pipeline
